include/meta/sequence/hmm/forward_backward.h

```diff
@@ -0,0 +1,385 @@
+            }
+        }
+
+        return bwd;
+    }
+
+    template <class HMM>
+    static util::dense_matrix<double>
+    posterior_state_membership(const HMM& hmm, const trellis& fwd,
+                               const trellis& bwd)
+    {
+        util::dense_matrix<double> gamma{fwd.size(), hmm.num_states()};
+        std::vector<double> scratch(hmm.num_states());
+        for (uint64_t t = 0; t < fwd.size(); ++t)
+        {
+            for (label_id i{0}; i < hmm.num_states(); ++i)
+            {
+                state_id s_i{i};
+                gamma(t, s_i) = fwd.probability(t, i) + bwd.probability(t, i);
+            }
+            auto norm = log_sum_exp(gamma.begin(t), gamma.end(t));
+            std::transform(gamma.begin(t), gamma.end(t), gamma.begin(t),
+                           [=](double val) { return val - norm; });
+        }
+        return gamma;
+    }
+
+    template <class HMM, class ExpectedCounts>
+    static void
+    increment_counts(const HMM& hmm, ExpectedCounts& counts,
+                     const typename HMM::sequence_type& seq, const trellis& fwd,
+                     const trellis& bwd,
+                     const util::dense_matrix<double>& log_gamma,
+                     const util::dense_matrix<double>& output_log_probs)
+    {
+        for (label_id i{0}; i < hmm.num_states(); ++i)
+        {
+            state_id s_i{i};
+
+            // add expected counts for initial state probabilities
+            counts.model_counts.increment_initial(s_i,
+                                                  std::exp(log_gamma(0, s_i)));
+
+            // add expected counts for transition probabilities
+            for (label_id j{0}; j < hmm.num_states(); ++j)
+            {
+                state_id s_j{j};
+
+                for (uint64_t t = 0; t < seq.size() - 1; ++t)
+                {
+                    auto log_xi_tij
+                        = log_gamma(t, s_i) + std::log(hmm.trans_prob(s_i, s_j))
+                          + output_log_probs(t + 1, s_j)
+                          + bwd.probability(t + 1, j) - bwd.probability(t, i);
+
+                    counts.model_counts.increment_transition(
+                        s_i, s_j, std::exp(log_xi_tij));
+                }
+            }
+
+            // add expected counts for observation probabilities
+            for (uint64_t t = 0; t < seq.size(); ++t)
+            {
+                counts.obs_counts.increment(seq[t], s_i,
+                                            std::exp(log_gamma(t, s_i)));
+            }
+        }
+
+        // compute contribution to the log likelihood
+        std::vector<double> scratch(hmm.num_states());
+        for (label_id i{0}; i < hmm.num_states(); ++i)
+        {
+            scratch[i] = fwd.probability(seq.size() - 1, i);
+        }
+        counts.log_likelihood += log_sum_exp(scratch.begin(), scratch.end());
+    }
+};
+}
+}
+}
+#endif
```

Comment:
I don't see an obvious way to consolidate the `logarithm` and `scaling` variants into the same code without introducing more complexity, so I guess we can just leave it as is.